theano/gof/opt.py

```diff
@@ -558,6 +614,39 @@ def process_node(self, fgraph, node):
         else:
             self.nodes_seen.add(node)
 
+    def get_merged_assert_input(self, node, candidate):
+        new_inputs = []
+        for node_i, cand_i in zip(node.inputs, candidate.inputs):
+            # if node_i is assert
+            if (node_i.owner and
+                    isinstance(node_i.owner.op,
+                               theano.tensor.opt.Assert)):
+                # node_i is assert, cand_i is assert
+                if (cand_i.owner and
+                        isinstance(cand_i.owner.op,
+                                   theano.tensor.opt.Assert)):
+                    # Here two assert nodes are merged.
+                    # Step 1. Check if two conditions the same one
+                    # Step 2. Combine the two with T.and_(a, b)
+                    node_cond = node_i.owner.inputs[1]
+                    cand_cond = cand_i.owner.inputs[1]
+                    if node_cond.owner is cand_cond.owner:
+                        new_inputs.append(cand_i)
+                    else:
+                        new_inputs.append(
+                            theano.tensor.opt.assert_op(
+                                node_i.owner.inputs[0],
+                                theano.tensor.and_(node_cond, cand_cond)))
```

Comment:
The order can be changed. Anyway there is no reason to put node before
candidate or the other way around.
Le 4 août 2015 04:40, "t13m" <notifications@github.com> a écrit :

> In theano/gof/opt.py
> <https://github.com/Theano/Theano/pull/3232#discussion_r36167717>:
>
> > +                # node_i is assert, cand_i is assert
> > +                if (cand_i.owner and
> > +                        isinstance(cand_i.owner.op,
> > +                                   theano.tensor.opt.Assert)):
> > +                    # Here two assert nodes are merged.
> > +                    # Step 1. Check if two conditions the same one
> > +                    # Step 2. Combine the two with T.and_(a, b)
> > +                    node_cond = node_i.owner.inputs[1]
> > +                    cand_cond = cand_i.owner.inputs[1]
> > +                    if node_cond.owner is cand_cond.owner:
> > +                        new_inputs.append(cand_i)
> > +                    else:
> > +                        new_inputs.append(
> > +                            theano.tensor.opt.assert_op(
> > +                                node_i.owner.inputs[0],
> > +                                theano.tensor.and_(node_cond, cand_cond)))
>
> Should the order of conditions be kept? I use new_cond =
> list(set(node_i.owner.inputs[1:] + cand_i.owner.inputs[1:])), so the
> order can be changed in the set.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/Theano/Theano/pull/3232/files#r36167717>.
>